src/D2L.CodeStyle.Analyzers/Immutability/TypeDeclarationImmutabilityAnalyzer.cs

```diff
@@ -100,26 +122,33 @@ INamedTypeSymbol typeSymbol
 			checker.CheckDeclaration( typeSymbol );
 		}
 
-		private static bool ShouldAnalyze(
-			INamedTypeSymbol analyzedType
+		private static void AnalyzeGenericMethodTypeArguments(
+			SyntaxNodeAnalysisContext ctx,
+			ImmutabilityContext immutabilityContext,
+			GenericNameSyntax syntax
 		) {
-			if( analyzedType.IsImplicitlyDeclared ) {
-				return false;
-			}
-
-			if( analyzedType.TypeKind == TypeKind.Interface ) {
-				return false;
-			}
-
-			if( Attributes.Objects.Immutable.IsDefined( analyzedType ) ) {
-				return true;
+			SymbolInfo info = ctx.SemanticModel.GetSymbolInfo( syntax, ctx.CancellationToken );
+			if( !( info.Symbol is IMethodSymbol method ) ) {
+				return;
 			}
 
-			if( Attributes.Objects.ImmutableBaseClass.IsDefined( analyzedType ) ) {
-				return true;
+			int i = 0;
+			var paramArgPairs = method.TypeParameters.Zip( method.TypeArguments, ( p, a ) => (p, a, i++) );
+			foreach( var (parameter, argument, position) in paramArgPairs ) {
+				if( !Attributes.Objects.Immutable.IsDefined( parameter ) ) {
+					continue;
+				}
+
+				if( !immutabilityContext.IsImmutable(
+					type: argument,
+					kind: ImmutableTypeKind.Total,

```

Comment:
could over time change this to check the value being passed to an argument (assuming the type arg is used for an argument) similar to checking initializers of members.

Reply:
Interesting. Open an issue? (Will forget about this personally)